src/graphics.rs

```diff
@@ -1628,6 +1634,263 @@ impl Drawable for Sprite {
+            .0
+            .pixel_bounding_box()
+            .unwrap_or(inval_bb)
+            .height() as usize;
+        let mut current_row: usize = 0;
+        let mut widest_row: usize = 0;
+        let mut row_sum: usize = 0;
+
+        // Sort the glyphs by height so that we know how tall each row should be in the atlas
+        // Sums all the widths and heights of the bounding boxes so we know how large the atlas will be
+        let mut char_rows = Vec::new();
+        let mut cur_row = Vec::with_capacity(chars_per_row);
+
+        for (glyph, c) in glyphs_and_chars.iter().rev() {
+            let bb = glyph.pixel_bounding_box().unwrap_or(inval_bb);
+
+            if current_row > chars_per_row {
+                current_row = 0;
+                texture_height += bb.height() as usize;
+                if row_sum > widest_row {
+                    widest_row = row_sum;
+                }
+                row_sum = 0;
+                char_rows.push(cur_row.clone());
+                cur_row.clear();
+            }
+
+            cur_row.push((glyph, *c));
+            row_sum += bb.width() as usize;
+            current_row += 1;
+        }
+        // Push remaining chars
+        char_rows.push(cur_row);
+
+        let texture_width = widest_row + (chars_per_row * MARGIN);
+        texture_height += chars_per_row * MARGIN;
+
+        let mut texture = RgbaImage::new(texture_width as u32, texture_height as u32);
+
+        let mut texture_cursor: (usize, usize) = (0, 0);
+
+        let mut char_map: HashMap<char, CharInfo> = HashMap::new();
+
+        for row in char_rows {
+            let height = row
+                .first()
+                .unwrap()
+                .0
+                .pixel_bounding_box()
+                .unwrap_or(inval_bb)
+                .height() as usize;
+            for (glyph, c) in row {
+                let bb = glyph.pixel_bounding_box().unwrap_or(inval_bb);
+
+                char_map.insert(
+                    c,
+                    CharInfo {
+                        vertical_offset: bb.min.y,
+                        uvs: Box2::new(
+                            texture_cursor.0 as f32 / texture_width as f32,
+                            texture_cursor.1 as f32 / texture_height as f32,
+                            bb.width() as f32 / texture_width as f32,
+                            bb.height() as f32 / texture_height as f32,
+                        ),
+                        width: glyph.unpositioned().h_metrics().advance_width,
+                    },
+                );
+
+                glyph.draw(|x, y, v| {
+                    let x: u32 = texture_cursor.0 as u32 + x;
+                    let y: u32 = texture_cursor.1 as u32 + y;
+                    let c = (v * 255.0) as u8;
+                    let color = Rgba([255, 255, 255, c]);
+                    texture.put_pixel(x, y, color);
+                });
+
+                texture_cursor.0 += bb.width() as usize + MARGIN;
```

Comment:
typically unless it's a pointer offset or a capacity I try to use `u32` for arbitrary unsigned integer values. most Rust crates follow this, which is why `bb.width()` returns a `u32` (iirc)